pkg/deploy/ec2/endpoint_service_manager.go

```diff
@@ -0,0 +1,265 @@
+
+	ReconcilePermissions(ctx context.Context, permissions *ec2model.VPCEndpointServicePermissions) error
+}
+
+// NewdefaultEndpointServiceManager constructs new defaultEndpointServiceManager.
+func NewDefaultEndpointServiceManager(ec2Client services.EC2, vpcID string, logger logr.Logger, trackingProvider tracking.Provider, taggingManager TaggingManager, externalManagedTags []string) *defaultEndpointServiceManager {
+	return &defaultEndpointServiceManager{
+		ec2Client:           ec2Client,
+		vpcID:               vpcID,
+		logger:              logger,
+		taggingManager:      taggingManager,
+		trackingProvider:    trackingProvider,
+		externalManagedTags: externalManagedTags,
+	}
+}
+
+var _ EndpointServiceManager = &defaultEndpointServiceManager{}
+
+// default implementation for EndpointServiceManager.
+type defaultEndpointServiceManager struct {
+	ec2Client           services.EC2
+	vpcID               string
+	logger              logr.Logger
+	taggingManager      TaggingManager
+	trackingProvider    tracking.Provider
+	externalManagedTags []string
+}
+
+func (m *defaultEndpointServiceManager) Create(ctx context.Context, resSG *ec2model.VPCEndpointService) (ec2model.VPCEndpointServiceStatus, error) {
+	sgTags := m.trackingProvider.ResourceTags(resSG.Stack(), resSG, resSG.Spec.Tags)
+	sdkTags := convertTagsToSDKTags(sgTags)
+
+	var resolvedLoadBalancerArns []string
+	for _, unresolved := range resSG.Spec.NetworkLoadBalancerArns {
+		arn, err := unresolved.Resolve(ctx)
+		if err != nil {
+			return ec2model.VPCEndpointServiceStatus{}, err
+		}
+		resolvedLoadBalancerArns = append(resolvedLoadBalancerArns, arn)
+	}
+
+	var privateDnsName *string
+	if resSG.Spec.PrivateDNSName != nil {
+		privateDnsName = awssdk.String(*resSG.Spec.PrivateDNSName)
+	}
+
+	req := ec2sdk.CreateVpcEndpointServiceConfigurationInput{
+		AcceptanceRequired:      awssdk.Bool(*resSG.Spec.AcceptanceRequired),
+		PrivateDnsName:          privateDnsName,
+		NetworkLoadBalancerArns: awssdk.StringSlice(resolvedLoadBalancerArns),
+		TagSpecifications: []*ec2sdk.TagSpecification{
+			{
+				ResourceType: awssdk.String("vpc-endpoint-service"),
+				Tags:         sdkTags,
+			},
+		},
+	}
+	m.logger.Info("creating VpcEndpointService",
+		"resourceID", resSG.ID())
+	resp, err := m.ec2Client.CreateVpcEndpointServiceConfigurationWithContext(ctx, &req)
+	if err != nil {
+		return ec2model.VPCEndpointServiceStatus{}, err
+	}
+	serviceID := awssdk.StringValue(resp.ServiceConfiguration.ServiceId)
+	m.logger.Info("created VpcEndpointService",
+		"resourceID", resSG.ID(),
+		"serviceID", serviceID)
+
+	return ec2model.VPCEndpointServiceStatus{
+		ServiceID: serviceID,
+	}, nil
+}
+
+func (m *defaultEndpointServiceManager) Update(ctx context.Context, resES *ec2model.VPCEndpointService, sdkES networking.VPCEndpointServiceInfo) (ec2model.VPCEndpointServiceStatus, error) {
+
+	m.logger.Info("Updating", "resES", resES, "sdkES", sdkES)
```

Comment:
Do we want to log this if no changes are being made?

Reply:
In `pkg/deploy/ec2/endpoint_service_synthesizer.go` we calculate the endpoint services that need created, deleted and updates and only call the relevant function.  If no endpoint services need updated then this function is never called.  It feels safe to me.